pkg/kubeutils/installation.go

```diff
@@ -111,6 +122,130 @@ func writeInstallationStatusMessage(writer *spinner.MessageWriter, install *ecv1
 	}
 }
 
+type RecordInstallationOptions struct {
+	IsAirgap           bool
+	Proxy              *ecv1beta1.ProxySpec
+	K0sConfig          *k0sv1beta1.ClusterConfig
+	License            *kotsv1beta1.License
+	ConfigSpec         *ecv1beta1.ConfigSpec
+	MetricsBaseURL     string
+	RuntimeConfig      *ecv1beta1.RuntimeConfigSpec
+	K0sConfigOverrides string
+}
+
+func RecordInstallation(ctx context.Context, kcli client.Client, opts RecordInstallationOptions) (*ecv1beta1.Installation, error) {
+	// ensure that the embedded-cluster namespace exists
+	ns := corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: runtimeconfig.EmbeddedClusterNamespace,
+		},
+	}
+	if err := kcli.Create(ctx, &ns); err != nil && !k8serrors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("create embedded-cluster namespace: %w", err)
+	}
+
+	// ensure that the installation CRD exists
+	if err := EnsureInstallationCRD(ctx, kcli); err != nil {
+		return nil, fmt.Errorf("create installation CRD: %w", err)
+	}
+
+	var euOverrides string
+	if opts.K0sConfigOverrides != "" {
+		eucfg, err := helpers.ParseEndUserConfig(opts.K0sConfigOverrides)
+		if err != nil {
+			return nil, fmt.Errorf("process overrides file: %w", err)
+		}
+		if eucfg != nil {
+			euOverrides = eucfg.Spec.UnsupportedOverrides.K0s
+		}
+	}
+
+	installation := &ecv1beta1.Installation{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: ecv1beta1.GroupVersion.String(),
+			Kind:       "Installation",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: time.Now().Format("20060102150405"),
+		},
+		Spec: ecv1beta1.InstallationSpec{
+			ClusterID:                 metrics.ClusterID().String(),
+			MetricsBaseURL:            opts.MetricsBaseURL,
+			AirGap:                    opts.IsAirgap,
+			Proxy:                     opts.Proxy,
+			Network:                   helpers.NetworkSpecFromK0sConfig(opts.K0sConfig),
+			Config:                    opts.ConfigSpec,
+			RuntimeConfig:             opts.RuntimeConfig,
+			EndUserK0sConfigOverrides: euOverrides,
+			BinaryName:                runtimeconfig.BinaryName(),
+			LicenseInfo: &ecv1beta1.LicenseInfo{
+				IsDisasterRecoverySupported: opts.License.Spec.IsDisasterRecoverySupported,
+				IsMultiNodeEnabled:          opts.License.Spec.IsEmbeddedClusterMultiNodeEnabled,
+			},
+		},
+	}
+	if err := CreateInstallation(ctx, kcli, installation); err != nil {
+		return nil, fmt.Errorf("create installation: %w", err)
+	}
+
+	// the kubernetes api does not allow us to set the state of an object when creating it
+	err := SetInstallationState(ctx, kcli, installation, ecv1beta1.InstallationStateKubernetesInstalled, "Kubernetes installed")
+	if err != nil {
+		return nil, fmt.Errorf("set installation state to KubernetesInstalled: %w", err)
+	}
+
+	return installation, nil
+}
+
+func EnsureInstallationCRD(ctx context.Context, kcli client.Client) error {
```

Comment:
moved this from `pkg/addons/embeddedclusteroperator/crds.go` so api and lam packages don't have to import the operator package